### PR TITLE
Variable-arity returns and function contracts.

### PR DESCRIPTION
**Pull request by Moderocky**
## Description

This provides better support for things that aren't able to determine their return arity during parse time.

Typically, things will error during parse time if you, for example, try to set a single variable `{var}` to several values `{list::*}`.
This is fine and intended behaviour, and helps to prevent a lot of silly mistakes (or people not realising that something returns multiple things), however it also [causes annoying issues](<https://github.com/SkriptLang/Skript/issues/6488>) with things that conceivably could be simple (and there are real cases where it should be permissible to `set {x} to ...`) but aren't able to determine whether they *will* be single.

Two features are added to assist with this.

### Function/Callable Contracts
This addresses the problem in the linked issue.
Some functions (currently only `clamp` that I could find, there might be others where this could be applied) are marked non-single but are guaranteed to return only a single value when a single input is used.
This means that it can never use a single variable (e.g. `set {x} to clamp(0, 1, 2)`) _even if_ it's a guarantee that the function will return only one value.

This feature adds optional function **contracts**. A contract is like an agreement for the return type hints/singularity/etc. _based on_ the actual argument expressions used.
For example, a function could be single if `expression[0].isSingle()` or it might always return the return type of `expression[n]`, or return the same number of values as argument (e.g. some kind of `packIntoList(A, B, C)` function).
Contracts allow us to specify this in order to have a more reliable signature and type hints at parsing time.  

This information is typically unavailable to a Function until it becomes a FunctionReference, and so wasn't really possible before.
Functions defer to their contract for return types and singularity, but a function reference can be its own contract so nothing has changed for most functions.

The `clamp` function has been changed to use a contract (single = arg 0 is single), fixing the linked issue. It will now be single with single inputs and properly error if used in the wrong way.

### Potential Singles
This is a user-knows-best model, and so should be **applied with caution**.
This is a special toggle for expressions that may return multiple values, but still ought to permit single changers in the `SET` mode.
If an expression returns true from `permitSingle()`, then it can be used as the change value for a single expression/variable in the `SET` mode.
E.g. `set {x} to blah` and `set {x::*} to blah` would **both** be permissible.
When used in single mode, the first value of the array is returned and everything else (if it exists) is discarded, which is actually what happens in a normal single expression.

This is designed for cases where singularity might be known only at runtime (or based on some factor Skript can't safely acquire) but the user probably has a good idea based on the use case. We **trust** the user to make the right decision.

Currently, no existing syntax uses this -- I felt it was out of scope for this PR.

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** #6488, alternative to #6567

### CODE REVIEW

**Review by sovdeeth — APPROVED** [review]
this looks very nice, exactly what i had wanted when I originally made clamp()!

File: src/main/java/ch/njol/skript/lang/Expression.java [hunk 1]
```diff
@@ -120,6 +120,20 @@ default Optional<T> getOptionalSingle(Event event) {
...
+	 * Whether there's a possibility this could return a single value.
+	 * Designed for expressions that may return more than one value, but are equally appropriate to use where
+	 * only singular values are accepted, in which case a single value (out of all available values) will be returned.
+	 * An example would be functions that return results based on their inputs.
+	 * Ideally, this will return {@link #isSingle()} based on its known inputs at initialisation, but for some syntax
+	 * this may not be known (or a syntax may be intentionally vague in its permissible returns).
+	 * @return Whether this can be used by single changers
+	 * @see #getSingle(Event)
+	 */
+	default boolean permitSingle() {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/lang/Expression.java:133]: I wonder if something like `mayBeSingle` would be a better name.

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/lang/Expression.java (hunk 1 above)
Comment (Moderocky) [reply src/main/java/ch/njol/skript/lang/Expression.java:133]: This was what I originally had, but I was concerned that the intended purpose isn't just for things that *could* be single but also for stuff that just wants to allow single use even if it isn't; a kind of "you can discard all other values and just have one if you like" thing.

File: src/main/java/org/skriptlang/skript/lang/function/Contract.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import ch.njol.skript.lang.Expression;
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * The 'contract' of a function or another callable.
+ * This is a non-exhaustive helper for type hints, singularity, etc. that may change based on the arguments
+ * passed to a callable, in order for it to make better judgements on correct use at parse time.
+ */
+public interface Contract {
```
[the remark this answers is not in the document]
Comment (Moderocky) [reply src/main/java/org/skriptlang/skript/lang/function/Contract.java:29]: Actually yes! I separated this because I felt it was something we might want to use in other fun places (like if we ever made another kind of functional _thing_, like a lambda or a subroutine section or some kind) or if it might pop up in something like skript-reflect too, since it's now our responsibility to take care of that.
Generally I make an effort to abstract any architecture I can so that we can reuse it in future as needed :)

File: src/main/java/org/skriptlang/skript/lang/function/Contract.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package org.skriptlang.skript.lang.function;
```
[the remark this answers is not in the document]
Comment (Moderocky) [reply src/main/java/org/skriptlang/skript/lang/function/Contract.java:19]: It didn't have anything to do with Bukkit so I thought you'd want it somewhere else, but sure I can move it if you prefer :)

File: src/main/java/org/skriptlang/skript/lang/function/Contract.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (APickledWalrus) [reply src/main/java/org/skriptlang/skript/lang/function/Contract.java:29]: It might be better to stick this in a util package then

File: src/main/java/ch/njol/skript/lang/Expression.java (hunk 1 above)
Comment (sovdeeth) [reply src/main/java/ch/njol/skript/lang/Expression.java:133]: I like canBeSingle

File: src/test/skript/tests/syntaxes/functions/clamp.sk [hunk 4]
```diff
@@ -41,3 +41,15 @@ test "clamp numbers":
 	assert number within {_got::1} is not number within {_got::1} with "(edge cases list) NaN" # need within because the variables weren't cooperating
 	assert {_got::2} is {_expected::2} with "(edge cases list) -infinity"
 	assert {_got::3} is {_expected::3} with "(edge cases list) infinity"
+
+test "clamp numbers (single)":
+	set {_expected::*} to (1, 0.0, and 2.0)
+	set {_got::*} to clamp((1, -infinity value, infinity value), 0.0, 2.0)
+	assert size of {_got::*} is 3 with "got %size of {_got::*}%"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/functions/clamp.sk:48]: you know what i'm here about

Comment (Moderocky) [reply src/test/skript/tests/syntaxes/functions/clamp.sk:48]: You're the grim reaper of my evening can you at least suggest the change so I don't have to open the IDE 😭

Comment (sovdeeth) [reply src/test/skript/tests/syntaxes/functions/clamp.sk]: ☠️

File: src/main/java/org/skriptlang/skript/lang/util/Contract.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+import ch.njol.skript.lang.Expression;
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * The 'contract' of a function or another callable.
+ * This is a non-exhaustive helper for type hints, singularity, etc. that may change based on the arguments
+ * passed to a callable, in order for it to make better judgements on correct use at parse time.
+ */
+public interface Contract {
```
Comment (APickledWalrus) [inline src/main/java/org/skriptlang/skript/lang/util/Contract.java:29]: I might prefer this in the ch.njol package due to its usage of ch.njol classes that may be replaced in the future

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/lang/Expression.java [hunk 6]
```diff
@@ -120,6 +120,20 @@ default Optional<T> getOptionalSingle(Event event) {
...
+	 * Whether there's a possibility this could return a single value.
+	 * Designed for expressions that may return more than one value, but are equally appropriate to use where
+	 * only singular values are accepted, in which case a single value (out of all available values) will be returned.
+	 * An example would be functions that return results based on their inputs.
+	 * Ideally, this will return {@link #isSingle()} based on its known inputs at initialisation, but for some syntax
+	 * this may not be known (or a syntax may be intentionally vague in its permissible returns).
+	 * @return Whether this can be used by single changers
+	 * @see #getSingle(Event)
+	 */
+	default boolean canBeSingle() {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/lang/Expression.java:133]: Since https://github.com/SkriptLang/Skript/pull/6624 uses `mayReturn`, I think we should decide between `canX` vs `mayX` for consistency.

File: src/main/java/ch/njol/skript/util/Contract.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+	 * @see Expression#isSingle()
+	 */
+	boolean isSingle(Expression<?>... arguments);
+
+	/**
+	 * @return What this will return, given these parameters
+	 * @see Expression#getReturnType()
+	 */
+	@Nullable
+	Class<?> getReturnType(Expression<?>... arguments);
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/util/Contract.java:42]: How might this interact with https://github.com/SkriptLang/Skript/pull/6624? Maybe we should consider having this method as `getReturnTypes`.

File: src/main/java/ch/njol/skript/lang/function/Signature.java [hunk 8]
```diff
@@ -126,6 +144,10 @@ public String getOriginClassPath() {
 		return originClassPath;
 	}
 
+	public Contract getContract() {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/lang/function/Signature.java:147]:
```suggestion
	@Nullable
	public Contract getContract() {
```

File: src/main/java/ch/njol/skript/util/Contract.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+import ch.njol.skript.lang.Expression;
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * The 'contract' of a function or another callable.
+ * This is a non-exhaustive helper for type hints, singularity, etc. that may change based on the arguments
+ * passed to a callable, in order for it to make better judgements on correct use at parse time.
+ */
+public interface Contract {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/util/Contract.java:29]: do you think this is worth generifying? mainly for return types.

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/util/Contract.java (hunk 7 above)
Comment (Moderocky) [reply src/main/java/ch/njol/skript/util/Contract.java:42]: So I see why you're saying that, but _probably not_.

If we _don't_ have a contract then we could conceivably have multiple possible return types, but with a contract we shouldn't:
The purpose of the contract is basically to short-circuit that kind of problem with functions: with a contract we know the _exact_ return type based on the inputs, so there shouldn't really ever be a multitude of possible return types.
(If there are, you're probably misusing the contract!)

Basically, contracts are kind of trying to avoid situations where we would normally use the multi-return type thingy by being more specific.

File: src/main/java/ch/njol/skript/util/Contract.java (hunk 9 above)
Comment (Moderocky) [reply src/main/java/ch/njol/skript/util/Contract.java:29]: We could do, but I'm not a huge fan of type parameters where they're not providing value.

I can't think of a situation where we would actually know the generic type of this and be able to use it _without an unchecked cast_.
Even if we got the type from the function and returned it (in the function call expression) we'd still need to perform an unchecked cast for the function return because with (for example) `Function<A>` and `Contract<B>` the scope of the reified parameters probably won't be the same.

File: src/main/java/ch/njol/skript/lang/Expression.java (hunk 6 above)
Comment (Moderocky) [reply src/main/java/ch/njol/skript/lang/Expression.java:133]: Originally #6624 didn't use `may` but that was a requested change there. Would you like both to be `may` or both to be `can` ?

Comment (sovdeeth) [reply src/main/java/ch/njol/skript/lang/Expression.java:133]: I think we should stick with canX, i can't find any other mayX's but i did find multiple canX's.

**Review by UnderscoreTud — APPROVED** [review]

**Review by APickledWalrus — APPROVED** [review]
